gradio/utils.py

```diff
@@ -346,3 +349,222 @@ def run_coro_in_background(func: Callable, *args, **kwargs):
+        (2) Since we're still using httpx.Request class by wrapping it, we have all it's functionalities.
+        (3) Exceptions are handled silently during the request call, which gives us the ability to inspect each one
+        individually in the case of multiple asynchronous request calls and some of them failing.
+        (4) Provides HTTP request types with Request.Method Enum class for ease of usage
+    Request also offers some util functions such as has_exception, is_valid and status to inspect get detailed
+    information about executed request call.
+
+    The basic usage of Request is as follows: create a Request object with inputs(method, url etc.). Then use it
+    with the "await" statement, and then you can use util functions to do some post request checks depending on your use-case.
+    Finally, call the get_validated_data function to get the response data.
+
+    You can see example usages in test_utils.py.
+    """
+
+    ResponseJson = NewType("ResponseJson", Json)
+
+    class Method(str, Enum):
+        """
+        Method is an enumeration class that contains possible types of HTTP request methods.
+        """
+
+        ANY = "*"
+        CONNECT = "CONNECT"
+        HEAD = "HEAD"
+        GET = "GET"
+        DELETE = "DELETE"
+        OPTIONS = "OPTIONS"
+        PATCH = "PATCH"
+        POST = "POST"
+        PUT = "PUT"
+        TRACE = "TRACE"
+
+    def __init__(
+        self,
+        method: Method,
+        url: str,
+        *,
+        validation_model: Type[BaseModel] = None,
+        validation_function: Callable = None,
+        exception_type: Type[Exception] = Exception,
+        raise_for_status: bool = False,
+        **kwargs,
+    ):
+        """
+        Initialize the Request instance.
+        Args:
+            method(Request.Method) : method of the request
+            url(str): url of the request
+            *
+            validation_model(Type[BaseModel]): a pydantic validation class type to use in validation of the response
+            validation_function(Callable): a callable instance to use in validation of the response
+            exception_class(Type[Exception]): a exception type to throw with its type
+            raise_for_status(bool): a flag that determines to raise httpx.Request.raise_for_status() exceptions.
+        """
+        self._response = None
+        self._exception = None
+        self._status = None
+        self._raise_for_status = raise_for_status
+        self._validation_model = validation_model
+        self._validation_function = validation_function
+        self._exception_type = exception_type
+        self._validated_data = None
+        # Create request
+        self._request = self._create_request(method, url, **kwargs)
+
+    def __await__(self):
+        """
+        Wrap Request's __await__ magic function to create request calls which are executed in one line.
+        """
+        return self.__run().__await__()
```

Comment:
I'm confused what's happening here. It looks like you are recursively calling the `__await__()` function?

Reply:
Yeah I was confused initially as well, but it was clear to me thx to @hibestil 

Request.await() ->
Runs
1. Request.init
2. Request.await
->
Runs
Request.__run

And later-on we await the async __run function.

Reply:
Sorry, still not clear to me.

In step (2), we run  `self.__run().__await__()`, but `self.__run()` returns the Request object, so are we not running Request.__run() again and again?

Reply:
Let me try again.

Awaiting a Request instance causes `__await__` function to run, which awaits the `__run` function. See [here](https://docs.python.org/3/reference/datamodel.html?highlight=__await__#awaitable-objects) for more

In Short
`await Request()`
1. creates a Request object
2. Awaits the __run function

Reply:
Thanks, @FarukOzderim for your clarifications. In addition to yours, I'd like to share the process step by step:
- When an instance is created using the Request class 
 ```python
 client_response = await Request(method=Request.Method.GET, url="http://headers.jsontest.com/")
```
- `await` statement calls the` Request.__await__` function to define the behavior of the `Request` instance when it is awaited.
- In  the` Request.__await__` function, we call the `Request.__run` function to execute the request stuff. 
    - Since `__run` is an async function, we have to await it, so here we used `self.__run().__await__()` to call an async function inside a non-async function.

**Briefly:** Using the `await` statement while the creation of the Request instance leads us to execute the `Request.__run` method. So we are not calling a thing recursively.


**Reference:**
https://docs.python.org/3/reference/datamodel.html#object.__await__